### PR TITLE
Support `uv add -r requirements.txt`

### PR DESCRIPTION
**Pull request by blueraft**
## Summary

Resolves https://github.com/astral-sh/uv/issues/4537

- First commit avoids overwriting dependencies with different markers.
- Second commit supports adding from requirements files.

## Test Plan

`cargo test`

### CODE REVIEW

File: crates/uv/src/commands/project/remove.rs [hunk 1]
```diff
@@ -93,29 +95,30 @@ pub(crate) async fn remove(
     }?;
 
     for package in packages {
+        let requirement = Requirement::from_str(package.as_str())?;
```
Comment (zanieb) [inline crates/uv/src/commands/project/remove.rs:98]: I don't think we should do this cast / change the removal APIs, why was this needed? Removal operates on package names, not full requirements.

File: crates/uv-cli/src/lib.rs [hunk 2]
```diff
@@ -2360,9 +2360,18 @@ pub struct LockArgs {
 #[derive(Args)]
 #[allow(clippy::struct_excessive_bools)]
 pub struct AddArgs {
-    /// The packages to add, as PEP 508 requirements (e.g., `ruff==0.5.0`).
-    #[arg(required = true)]
-    pub requirements: Vec<String>,
+    /// Install all listed packages.
```
Comment (zanieb) [inline crates/uv-cli/src/lib.rs:2363]: Not quite "Install" here, can we retain the previous description? It'd be good to move the "as PEP 508" part to a second line though so it's only present in the long help.

File: crates/uv-cli/src/lib.rs [hunk 3]
```diff
@@ -2360,9 +2360,18 @@ pub struct LockArgs {
...
+    #[arg(group = "sources")]
+    pub packages: Vec<String>,
+
+    /// Install all packages listed in the given `requirements.txt` files.
+    ///
+    /// If a `pyproject.toml`, `setup.py`, or `setup.cfg` file is provided, uv will
+    /// extract the requirements for the relevant project.
```
Comment (zanieb) [inline crates/uv-cli/src/lib.rs:2370]: Should we allow these files? These seem like a bit of a misuse. We don't allow them in `uv run` https://github.com/astral-sh/uv/blob/2822dde8cb24f1486cef59228c723da931690c30/crates/uv/src/commands/project/run.rs#L74-L92

File: crates/uv-cli/src/lib.rs [hunk 4]
```diff
@@ -2360,9 +2360,18 @@ pub struct LockArgs {
...
 #[allow(clippy::struct_excessive_bools)]
 pub struct AddArgs {
-    /// The packages to add, as PEP 508 requirements (e.g., `ruff==0.5.0`).
-    #[arg(required = true)]
-    pub requirements: Vec<String>,
+    /// Install all listed packages.
+    #[arg(group = "sources")]
+    pub packages: Vec<String>,
+
+    /// Install all packages listed in the given `requirements.txt` files.
```
Comment (zanieb) [inline crates/uv-cli/src/lib.rs:2367]: Same as above, not "install".

**Comment by zanieb** [comment]
Could you separate the change to avoid marker collisions from the `requirements.txt` support? I think they'll make more sense to discuss in two pull requests.

File: crates/uv/src/commands/project/remove.rs (hunk 1 above)
Comment (zanieb) [reply crates/uv/src/commands/project/remove.rs:98]: I see this probably is related to https://github.com/astral-sh/uv/pull/6005/commits/0fbde82c7dbe1ac9fd40919d642f5976d2b79d08 now, perhaps we can cast to a `Requirement` later instead if needed?

File: crates/uv-cli/src/lib.rs (hunk 3 above)
Comment (blueraft) [reply crates/uv-cli/src/lib.rs:2372]: Supporting `setup.py` and `setup.cfg` files could be a helpful feature, making it easier for users to transition to the pyproject.toml standard.

**Comment by charliermarsh** [comment]
Assigning myself.

File: crates/uv-cli/src/lib.rs (hunk 3 above)
Comment (charliermarsh) [reply crates/uv-cli/src/lib.rs:2372]: It's possible that this could be a killer feature because you could import Poetry projects this way... But it doesn't actually work as-is. I'm gonna remove it and look at it in a separate PR.

**Review by charliermarsh — APPROVED** [review]
